pgx/src/datum/array.rs

```diff
@@ -99,22 +99,13 @@ impl<'a, T: FromDatum + serde::Serialize> serde::Serialize for Array<'a, T> {
 
 impl<'a, T: FromDatum> Drop for Array<'a, T> {
     fn drop(&mut self) {
-        if let Array { raw, datum_palloc: Some(data), elem_slice, .. } = self {
-            // If Drop has arrived here, it means that this Array is backed by an allocation or two
-            // If so, the first one is guaranteed, and was created by calling pg_sys::deconstruct_array
-            // This is just a slice, dropping it doesn't "do" anything, but out of an abundance of caution:
-            mem::drop(elem_slice);
-            // Now there shouldn't be any other references to that slice, so this can be deallocated:
-            unsafe { pg_sys::pfree(data.as_ptr().cast()) };
-
-            // Detoasting the varlena may have allocated: the toasted varlena cloned as a detoasted ArrayType
-            // Checking for pointer equivalence is the only way we can truly tell
-            let raw = raw.take().map(|r| r.into_ptr());
-            if let Some(raw) = raw {
+        // First drop the slice that references the RawArray
+        let slice = mem::take(&mut self.datum_slice);
+        mem::drop(slice);
+        if self.needs_pfree {
+            if let Some(raw) = self.raw.take().map(|r| r.into_ptr()) {
```

Comment:
What's the point of `self.needs_pfree` then?

Reply:
It indicates the "was cloned on detoasting" case, the `RawArray` resides in an `Option<_>` to allow `mem::take` on it (basically, Advanced Rust Antics), this `if let` will always pass. I have been thinking maybe I should use something more like the pattern for `Cow` here, but I want to take another shot, later, at removing the need for `Option` first before embracing this as code that will remain here.